### PR TITLE
Sign simulator bundles with com.apple.security.get-task-allow

### PR DESCRIPTION
**Pull request by brentleyjones**
This is needed for Accessibility Inspector to work.

### CODE REVIEW

**Comment by kastiglione** [comment]
Thanks for finding and fixing this!

File: apple/internal/entitlement_rules.bzl [hunk 1]
```diff
@@ -281,13 +281,37 @@ def _entitlements_impl(ctx):
...
+            )
+
         return [
             linking_support.sectcreate_objc_provider(
                 "__TEXT",
                 "__entitlements",
                 final_entitlements,
             ),
-            AppleEntitlementsInfo(final_entitlements = final_entitlements),
+            AppleEntitlementsInfo(final_entitlements = simulator_entitlements),
```
Comment (keith) [inline apple/internal/entitlement_rules.bzl:314]: Do we not need to also propagate the previous `final_entitlements` somehow here?

Comment (brentleyjones) [reply apple/internal/entitlement_rules.bzl:314]: I don't believe so. the way Xcode does this is embed these entitlements while linking (which we do right above) and only signed with `com.apple.security.get-task-allow`. `AppleEntitlementsInfo.final_entitlements` is what Bazel signs with, so we only want to propagate `simulator_entitlements` here. We use `final_entitlements` in `sectcreate_objc_provider`, so it doesn't get dropped as unused.

**Comment by brentleyjones** [comment]
Rebased

**Comment by keith** [comment]
looks like the test failures are real and the API you're calling has drifted a bit. would also be nice to cover this in a small test case if possible

**Comment by brentleyjones** [comment]
I'll ping when I've added a test.

**Comment by thomasvl** [comment]
Drive by - Should this replace https://github.com/bazelbuild/rules_apple/blob/master/doc/common_info.md#debugging-entitlement-support-appleadd_debugger_entitlement?  i.e. - it was meant to add this when needed, if you always force it for simulator builds, does the flag need to exist for device builds?  (or should the existing logic that honors that also check if it is a simulator build and force it there?)

**Comment by brentleyjones** [comment]
This is on top of that, since even with "The rules will auto add the entitlement for non optimized builds", which results in linker sections for the simulator, we _also_ need this one flag in the entitlements. This is still gated behind `_include_debug_entitlements`, so we will only do this under the same situations that you had laid out.

**Comment by thomasvl** [comment]
If you nest a device/simulator check in 268-268 and also add this to the force list at that time, you don't need to write out any other files, and the existing action flow should do all the work, no?  Should be a much shorter edit and doesn't require extra actions.

**Comment by brentleyjones** [comment]
This new entitlement can't have anything but thing flag, if i recall correctly. If we sign it with the same entitlement that goes in the linking then the Accessibility Inspector is still broken. I'll confirm this again as I fix this up and add tests.

File: test/ios_application_test.sh [hunk 2]
```diff
@@ -351,12 +353,26 @@ EOF
 
     readonly FILE_TO_CHECK="${TEST_TMPDIR}/dumped_entitlements"
     readonly SHOULD_CONTAIN="${FOR_SIM}"
+
+    # Simulator builds also have entitlements in the codesign output,
+    # but only `com.apple.security.get-task-allow` and nothing else
+    do_build ios "$@" //app:dump_codesign || fail "Should build"
```
Comment (brentleyjones) [inline test/ios_application_test.sh:359]: I couldn't make the do_build on 350 to build this as well. Open to a change that make it only build once though.

File: test/ios_application_test.sh [hunk 3]
```diff
@@ -351,12 +353,26 @@ EOF
...
+      assert_not_contains "<key>keychain-access-groups</key>" "${CODESIGN_OUTPUT}"
+    fi
   fi
 
   if [[ "${SHOULD_CONTAIN}" == "y" ]] ; then
     assert_contains "<key>get-task-allow</key>" "${FILE_TO_CHECK}"
+    assert_contains "<key>keychain-access-groups</key>" "${FILE_TO_CHECK}"
   else
     assert_not_contains "<key>get-task-allow</key>" "${FILE_TO_CHECK}"
+    assert_contains "<key>keychain-access-groups</key>" "${FILE_TO_CHECK}"
```
Comment (brentleyjones) [inline test/ios_application_test.sh:375]: Added these to verify that the existing entitlements are included with the defaults.

File: test/ios_application_test.sh [hunk 4]
```diff
@@ -351,12 +353,26 @@ EOF
...
+    # Simulator builds also have entitlements in the codesign output,
+    # but only `com.apple.security.get-task-allow` and nothing else
+    do_build ios "$@" //app:dump_codesign || fail "Should build"
+
+    if [[ "${SHOULD_CONTAIN}" == "y" ]] ; then
+      assert_contains "<key>com.apple.security.get-task-allow</key>" "${CODESIGN_OUTPUT}"
+      assert_not_contains "<key>keychain-access-groups</key>" "${CODESIGN_OUTPUT}"
+    else
+      assert_not_contains "<key>com.apple.security.get-task-allow</key>" "${CODESIGN_OUTPUT}"
+      assert_not_contains "<key>keychain-access-groups</key>" "${CODESIGN_OUTPUT}"
```
Comment (brentleyjones) [inline test/ios_application_test.sh:366]: This specifically tests that we are only getting the default, and not the full entitlements. This is how Xcode does it, so even if it's not a requirement I would prefer to match that.

**Review by brentleyjones — COMMENTED** [review]
@keith Tests fixed and augmented to cover this use case.

**Review by keith — APPROVED** [review]

**Pull request merged**